Improve Gen UI docs, group providers/gen ui in API reference, misc improvements (#1014)
vercel/ai: docs/pages/docs/api-reference/generative-ui/render.mdx

---
title: render
---

import { Callout, Tabs, Tab, Steps } from 'nextra-theme-docs';
import { UIPreviewCard, Card } from '@/components/home/card';
import { Browser } from '@/components/home/browser';
import { EventPlanning } from '@/components/home/event-planning';
import { Searching } from '@/components/home/searching';

# render()

> The `render` function is a helper function to create a streamable UI from an LLM response. It currently only supports

Reviewer: ```suggestion
The `render` function is a helper function to create a streamable UI from an LLM response. It currently only supports OpenAI-compatible models/providers. By default, the `text` method will return a React fragment wrapped around the text content of the LLM response. 

The `render` function allows you to map [OpenAI-compatible Tool/Function Calls](https://platform.openai.com/docs/guides/function-calling) to [React Server Components](https://vercel.com/blog/understanding-react-server-components) using the `tools` key. Both `text` and `render` can be specified, and `text` being the fallback for when no function is called by the model.

> OpenAI API compatible models with function calling, like gpt-3.5-turbo and [FireFunction](https://fireworks.ai/blog/firefunction-v1-gpt-4-level-function-calling).

Reviewer 2: Currently the `provider` is OpenAI only.

> OpenAI API compatible models with function calling, like gpt-3.5-turbo and [FireFunction](https://fireworks.ai/blog/firefunction-v1-gpt-4-level-function-calling).

Reviewer: ```suggestion
```

> `render` provides a generator-based API for streaming updating UIs from the LLM.

Reviewer 2: If fact, the “render” function can be one of the four:
- () => ReactNode
- async () => ReactNode
- function* () {}
- async function* () {}

> `render` provides a generator-based API for streaming updating UIs from the LLM.

Reviewer: ```suggestion
`render` provides a few APIs for streaming updating UIs from the LLM:
- `() => ReactNode` - a function that returns a React Node
- `async () => ReactNode` - an async function that returns a React Node
- `function* () {}` - an generator function that returns a React Node.
- `async function* () {}` - an async generator function that returns a React Node.
```

              ...aiState.get(),
              {
                role: "assistant",
                content
              }
            })
          }

          return <div>{content}</div>
        }
        tools: {
          get_flight_info: {
            description: 'Get the information for a flight',
            parameters: z.object({
              flightNumber: z.string().describe('the number of the flight')
            }).required(),
            // flightNumber is inferred from the parameters passed above
            render: async function* ({ flightNumber }) {
              yield <Spinner/>
              const flightInfo = await getFlightInfo(flightNumber)

              aiState.done([
                ...aiState.get(),
                {
                  role: "function",
                  name: "get_flight_info",
                  // Content can be any string to provide context to the LLM in the rest of the conversation
                  content: JSON.stringify(flightInfo),
                }
              ]);

              return <FlightCard flightInfo={flightInfo} />
            }
          }
        }
      })

      return {
        id: Date.now(),

>         // You can render ui on the client with something like `{message.ui}` and the

Reviewer 2: ```suggestion
        // You can render ui on the client with something like `{message.display}` and the
```
